pkg/metadata/service_discovery.go

```diff
@@ -15,12 +15,56 @@
 package metadata
 
 import (
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+
 	"github.com/parca-dev/parca-agent/pkg/discovery"
 )
 
 // ServiceDiscovery metadata provider.
-func ServiceDiscovery(m *discovery.Manager) *Provider {
-	return &Provider{
-		"service discovery", m.ProcessLabels,
+func ServiceDiscovery(logger log.Logger, m *discovery.Manager) Provider {
+	provider := &StatefulProvider{
+		name:  "service discovery",
+		state: map[int]model.LabelSet{},
+		mtx:   &sync.RWMutex{},
 	}
+
+	go func() {
+		defer level.Warn(logger).Log("msg", "service discovery metadata provider exited")
+
+		for tSets := range m.SyncCh() {
+			state := map[int]model.LabelSet{}
+			// Update process labels.
+			for _, groups := range tSets {
+				for _, group := range groups {
+					for _, pid := range group.PIDs {
+						// Overwrite the information we have here with the latest.
+						allLabels := model.LabelSet{}
+						for k, v := range group.Labels {
+							allLabels[k] = v
+						}
+						for _, t := range group.Targets {
+							for k, v := range t {
+								allLabels[k] = v
+							}
+						}
+
+						_, ok := state[pid]
+						if ok {
+							state[pid] = state[pid].Merge(allLabels)
+						} else {
+							state[pid] = allLabels
```

Comment:
This is the new concept I've introduced here. We don't cache the labels from service discovery since it manages its own state. That's why I have added this new method to the `Provider` interface.

Reply:
I like the state it is much simpler than a cache, but when the state is updated the labels cached by the label manager need to be invalidated

Reply:
I wasn't super concerned about the very front cache since the TTL was low. If you want, We can add a timestamp for the stateful provider for the last update time and check if it's changed, and invalidate any related cache in the label manager.